src/main/java/org/neo4j/junit/jupiter/causal_cluster/CausalCluster.java

```diff
@@ -18,44 +18,52 @@
  */
 package org.neo4j.junit.jupiter.causal_cluster;
 
-import java.net.URI;
-import java.util.Collection;
-
 import org.junit.jupiter.api.extension.ExtensionContext.Store.CloseableResource;
-import org.testcontainers.containers.GenericContainer;
-import org.testcontainers.containers.Neo4jContainer;
 import org.testcontainers.containers.SocatContainer;
 
+import java.net.URI;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static java.util.stream.Collectors.toList;
+
 /**
  * This allows us to interact with the Neo4j Causal Cluster
  *
  * @author Michael J. Simons
  */
-class CausalCluster implements CloseableResource {
+class CausalCluster implements Cluster, CloseableResource {
 
 	private final SocatContainer boltProxy;
-	private final Collection<Neo4jContainer> clusterMembers;
-	private final int boltPort;
+	private final List<Neo4jCore> clusterCores;
 
-	CausalCluster(SocatContainer boltProxy, int boltPort, Collection<Neo4jContainer> clusterMembers) {
+	CausalCluster(SocatContainer boltProxy, List<Integer> boltPorts, List<Neo4jCore> clusterCores) {
 
 		this.boltProxy = boltProxy;
-		this.clusterMembers = clusterMembers;
-		this.boltPort = boltPort;
+		this.clusterCores = clusterCores;
 	}
 
 	public URI getURI() {
-		return URI.create(String.format(
-			"neo4j://%s:%d",
-			boltProxy.getContainerIpAddress(),
-			boltProxy.getMappedPort(boltPort)
-		));
+		// Choose a random bolt port from the available ports
+		Neo4jCore core = clusterCores.get(ThreadLocalRandom.current().nextInt(0, clusterCores.size()));
```

Comment:
This is a bit opinionated on my part - I think it's better to return a random address rather than the first one every time.